field/TinyLang/Field/Generator.hs

```diff
@@ -396,7 +457,9 @@ instance (Field f, Arbitrary f) => Arbitrary (Statement f) where
     -- (which most of the time will break the assertion) we should shrink @lhs == rhs@ to
     -- @lhs' == lhs'@ or @rhs' == rhs'@ where @lhs'@ and @rhs'@ are shrunk version of
     -- @lhs@ and @rhs@ respectively (just to have some shrinking that does not break the assertion).
-    shrink (EAssert expr)    = EAssert <$> shrink expr
+    shrink (EAssert expr) = EAssert <$> shrink expr
+    shrink (EFor uniVar start end stmts) =  efor <$> shrink (start, end, stmts) where
```

Comment:
Shrinking `start` makes the loop more expensive to compute... Seems like if we shrink `start` to `start'`, then we should also subtract `start - start'` from `end`.

Reply:
Hmm, I need to think how would that work in the presence of permitting normalisable expressions in the for loop.

Reply:
Not necessarily, see my comment regarding weird loop ranges.

Reply:
I do not think this is a problem now.  I made a note, we can revisit it we notice that there are massive performance degradations for running tests.